Fix replace-type for different packages from the same source

RangelReale opened the pull request:
Description
-------------

When different packages that map to the same source package and mapped with replace-type, the package of the first one are used for all other packages.

- Fixes #688

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

Version of Golang used when building/testing:
---------------------------------------------

- [ ] 1.11
- [ ] 1.12
- [ ] 1.13
- [ ] 1.14
- [ ] 1.15
- [ ] 1.16
- [ ] 1.17
- [ ] 1.18
- [ ] 1.19
- [ ] 1.20
- [x] 1.21

How Has This Been Tested?
---------------------------

- a test was added for this specific problem.
- @rkoehl05 was able to validate the solution in his mock repo.

Checklist
-----------

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

RangelReale commented:
@LandonTClipp had to add the type to the `addPackageImportWithName` to be able to differentiate between them, I only replace if this parameter is not nil. Also added some specifity rule to check replaces with types first.

@rkoehl05 can you see if this fixes your problem?

rkoehl05 commented:
Thanks @RangelReale! I tested with my reproducer and I was able to generate a working mock.

LandonTClipp commented:
@RangelReale thanks so much for this. I converted this to a draft because you mentioned this is still a WIP. Please ping me when you're ready for me to review.

RangelReale commented:
> @RangelReale thanks so much for this. I converted this to a draft because you mentioned this is still a WIP. Please ping me when you're ready for me to review.

@LandonTClipp I updated the description, it should be ready now.

LandonTClipp commented on pkg/fixtures/example_project/fiz/fiz.go:
```diff
@@ -0,0 +1,11 @@
+type Fiz interface {
+	FA(f fiz1.Fiz1)
+	FB(f fiz2.Fiz2)
+}
```
I would request you add a short README.md to `example_project/fiz` explaining what this package is testing.

Also I'd like it to be renamed to something more descriptive like `replaceType` or something.

RangelReale replied:
renamed the files and types, and added a README.

LandonTClipp commented on pkg/generator.go:
```diff
@@ -173,23 +173,28 @@ func (g *Generator) getPackageScopedType(ctx context.Context, o *types.TypeName)
-		if !f(item.from, item.to) {
-			break
+	// check most specific first
+	for _, hasType := range []bool{true, false} {
```
Good solution 👍🏻

LandonTClipp reviewed (changes requested):
Good implementation. Just minor nitpicks about the naming that could make this a bit clearer.

LandonTClipp reviewed (approved):
Awesome thank you!

LandonTClipp commented:
@RangelReale if you could squash the commits down to one, we can merge away.

RangelReale commented:
> @RangelReale if you could squash the commits down to one, we can merge away.

rebased to master, see if this is ok.

LandonTClipp commented:
Thank you very much for getting this out. I lean heavily on people submitting PRs ❤️
